samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java

```diff
@@ -18,47 +18,178 @@
 import org.apache.samza.task.StreamTaskFactory;
 import org.apache.samza.util.Util;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
 
 /**
- * StreamProcessor can be embedded in any application or executed in a distributed environment (aka cluster) as an
- * independent process.
  * <p>
- * <b>Usage Example:</b>
- * <pre>
- * StreamProcessor processor = new StreamProcessor(1, config);
- * processor.start();
- * try {
- *  boolean status = processor.awaitStart(TIMEOUT_MS);    // Optional - blocking call
- *  if (!status) {
- *    // Timed out
- *  }
- *  ...
- * } catch (InterruptedException ie) {
- *   ...
- * } finally {
- *   processor.stop();
- * }
- * </pre>
- * Note: A single JVM can create multiple StreamProcessor instances. It is safe to create StreamProcessor instances in
+ *   StreamProcessor can be embedded in any application or executed in a distributed environment (aka cluster) as an
+ *   independent process.
+ * </p>
+ *
+ * <b>Note</b>: A single JVM can create multiple StreamProcessor instances. It is safe to create StreamProcessor instances in
  * multiple threads.
  */
 @InterfaceStability.Evolving
 public class StreamProcessor {
+  private static final Logger LOGGER = LoggerFactory.getLogger(StreamProcessor.class);
+
   private final JobCoordinator jobCoordinator;
-  private final StreamProcessorLifecycleListener lifecycleListener;
-  private final String processorId;
+  private final StreamProcessorLifecycleListener processorListener;
+  private final Object taskFactory;
+  private final Map<String, MetricsReporter> customMetricsReporter;
+  private final Config config;
+  private final long taskShutdownMs;
+  private ExecutorService executorService;
+  private volatile SamzaContainer container = null;
+
+  @VisibleForTesting
+  JobCoordinatorListener jobCoordinatorListener = null;
+
+  // Latch used to synchronize between the JobCoordinator thread and the container thread, when the container is
+  // stopped due to re-balancing
+  private volatile CountDownLatch jcContainerLatch = new CountDownLatch(1);
+  private volatile boolean processorOnStartCalled = false;
+
+  SamzaContainer createSamzaContainer(ContainerModel containerModel, int maxChangelogStreamPartitions, JmxServer jmxServer) {
+    return SamzaContainer.apply(
+        containerModel,
+        config,
+        maxChangelogStreamPartitions,
+        jmxServer,
+        Util.<String, MetricsReporter>javaMapAsScalaMap(customMetricsReporter),
+        taskFactory);
+  }
+
+  JobCoordinatorListener createJobCoordinatorListener() {
+    return new JobCoordinatorListener() {
+
+      @Override
+      public void onJobModelExpired() {
+        if (container != null && container.getStatus().equals(SamzaContainerStatus.RUNNING)) {
+          container.shutdown(true);
```

Comment:
Spoke with @xinyuiscool and @prateekm offline. Summarizing it here:
Above, Xinyu was suggesting the overwriting the listener to the container depending on who invokes the shutdown. Or register more than one listener to the container and the container invokes the correct listener depending on who is requesting the shutdown. 
The problem with either of the suggestion is:
1. we are blurring the contracts of the listener interface
2. container needs to be aware of the components it is interacting with. This is against the whole idea behind listener model

The main issue under discussion is the "pauseByJm" boolean flag passed to onContainerStop. This is required to identify why the container stopped or completed. If the container itself can clearly expose its state, then we wouldn't need this. Problem is that when the container "completes" processing (as in bounded or end of stream), it will automatically stop. There is no way to distinguish this state of the container from the "STOPPED" state. If we can clearly identify a "PAUSED" state and "COMPLETED" state of the SamzaContainer, we can get rid of this flag. Until such a clear definition of state becomes apparent, we have to thread the flag through the components. 

**Conclusion**: For the sake of clarity of the code, we decided to use an "enum" to identify the source requesting the shutdown, instead of using a boolean flag.